### ISSUE TITLE
activity type 6 (something similar happening again)

### ISSUE THREAD

**Issue by GameParrot** [JaxkDev/DiscordBot#120]
No ability to reconnect after the discord thread crashes makes it pretty much useless in production...

**Comment by JaxkDev** [comment]
Reconnecting to gateway from non-erroneous problems is supported (eg network instability etc) but if its crashing I wouldn’t want it to restart because theres an issue.

**Comment by GameParrot** [comment]
> Reconnecting to gateway from non-erroneous problems is supported (eg network instability etc) but if its crashing I wouldn’t want it to restart because theres an issue.

How do I do that?
Also its suddenly crashing within a few minutes after startup with no error in the log file

**Comment by JaxkDev** [comment]
> > Reconnecting to gateway from non-erroneous problems is supported (eg network instability etc) but if its crashing I wouldn’t want it to restart because theres an issue.
> 
> How do I do that? Also its suddenly crashing within a few minutes after startup with no error in the log file

It’s automatically done, and its almost never no reason in any log.
Check all log files both server.log and plugin_data logs

**Comment by GameParrot** [comment]
> > > Reconnecting to gateway from non-erroneous problems is supported (eg network instability etc) but if its crashing I wouldn’t want it to restart because theres an issue.
> > 
> > 
> > How do I do that? Also its suddenly crashing within a few minutes after startup with no error in the log file
> 
> It’s automatically done, and its almost never no reason in any log. Check all log files both server.log and plugin_data logs

No reason in any log, double checked
Running multiple instances results in disconnect from both at same time
I would recommend removing from poggit until this disconnect is fixed

**Comment by JaxkDev** [comment]
> > > > Reconnecting to gateway from non-erroneous problems is supported (eg network instability etc) but if its crashing I wouldn’t want it to restart because theres an issue.
> > > 
> > > 
> > > How do I do that? Also its suddenly crashing within a few minutes after startup with no error in the log file
> > 
> > 
> > It’s automatically done, and its almost never no reason in any log. Check all log files both server.log and plugin_data logs
> 
> No reason in any log, double checked Running multiple instances results in disconnect from both at same time I would recommend removing from poggit until this disconnect is fixed

Without concrete evidence this effects ‘majority’ of users I wont pull.

If you want me to even investigate your disconnections I needs logs and data, none of which I have from you.

Software versions
Hardware architecture 
Log files
Discord config (REDACT token)

**Comment by GameParrot** [comment]
Happens with latest pocketmine, Linux mint 21 (based on ubuntu 22.04), php 8.2.13 (happens also with 8.1), amd 5900x cpu (x64)
```
---
version: 4

type: "internal"

logging:
  max_files: 28
  directory: "logs"

protocol:
  general:
    packets_per_tick: 50
    heartbeat_allowance: 60

  internal:
    token: "---"

  external:
    host: "0.0.0.0"
    port: 22222

...
```
[DiscordBot-2023-12-23.log](https://github.com/DiscordBot-PMMP/DiscordBot/files/13759884/DiscordBot-2023-12-23.log)

**Comment by JaxkDev** [comment]
Looks like a graceful shutdown lets see serverlog

**Comment by JaxkDev** [comment]
I shall spin up a vm and see what’s happening locally for me as well.

**Comment by GameParrot** [comment]
> Looks like a graceful shutdown lets see serverlog

nothing there except for 
```
[13:11:32.645] [Server thread/NOTICE]: [DiscordBot] DiscordBot Disconnected, API no longer ready.
[13:11:32.645] [Server thread/ERROR]: [DiscordBot] Discord thread terminated, check logs for more information.
```

**Comment by JaxkDev** [comment]
whats before that, any events, chat, discord actions etc
Any other plugins using DiscordBot?

**Comment by GameParrot** [comment]
> Any other plugins using DiscordBot?

I'm using a plugin with it, but the crash is not after any event
Also it's really random when it will happen, like sometimes it takes under a minute and sometimes it doesn't happen at all
I think it could be a discord api change

**Comment by JaxkDev** [comment]
> > Any other plugins using DiscordBot?
> 
> I'm using a plugin with it, but the crash is not after any event Also it's really random when it will happen, like sometimes it takes under a minute and sometimes it doesn't happen at all I think it could be a discord api change

doubt, I have certain assertions and anything from discord would flag somewhere and throw an error.

If its random then its most likely an event causing it, but without knowing what happens it'll be close to impossible to say without getting some more info.

**Comment by GameParrot** [comment]
> doubt, I have certain assertions and anything from discord would flag somewhere and throw an error.

i can unload all plugins except for the discord plugin and test

**Comment by JaxkDev** [comment]
That could be helpful, just make sure its not external api requests being mishandled

**Comment by GameParrot** [comment]
> That could be helpful, just make sure its not external api requests being mishandled

Can confirm crash after a few minutes with no other plugins (only the discord bot plugin on empty server)

**Comment by JaxkDev** [comment]
I cant reproduce that, mines still running

**Comment by GameParrot** [comment]
> I cant reproduce that

Can i send you my discord token in dms
i created a bot just for testing

**Comment by JaxkDev** [comment]
> > I cant reproduce that
> 
> Can i send you my discord token in dms i created a bot just for testing

You may as long you understand the risks outlined by sharing a token, DO NOT add this test bot to ANY live server.
DO RESET THE TOKEN AFTER TESTING IS COMPLETE to avoid any mistaken invites.

**Comment by JaxkDev** [comment]
@JaxkDev

**Comment by GameParrot** [comment]
> > > I cant reproduce that
> > 
> > 
> > Can i send you my discord token in dms i created a bot just for testing
> 
> You may as long you understand the risks outlined by sharing a token, DO NOT add this test bot to ANY live server. DO RESET THE TOKEN AFTER TESTING IS COMPLETE to avoid any mistaken invites.

i dont think the crash will happen if its not in a live server

**Comment by JaxkDev** [comment]
> > > > I cant reproduce that
> > > 
> > > 
> > > Can i send you my discord token in dms i created a bot just for testing
> > 
> > 
> > You may as long you understand the risks outlined by sharing a token, DO NOT add this test bot to ANY live server. DO RESET THE TOKEN AFTER TESTING IS COMPLETE to avoid any mistaken invites.
> 
> i dont think the crash will happen if its not in a live server

in this case, remove ALL ADMIN permissions from bot and ensure it only has permission to do what you would be happy for it to do in the case this token is leaked.

**Comment by JaxkDev** [comment]
TLDR; `register_shutdown_function` in thread is being called with no error data from `error_get_last`

**Comment by JaxkDev** [comment]
New undocumented Hang Status activity type 6 (supposedly replacing VC status from what others can tell), leaving it disabled for now until official docs come.

The problem was ActivityType::from() was failing and throwing ValueError that somehow causes a fatal error to be thrown.

Resolved in v3.0.4
